cpp/src/arrow/compute/kernels/scalar_validity.cc

```diff
@@ -76,7 +79,26 @@ struct IsInfOperator {
 
 struct IsNullOperator {
   static Status Call(KernelContext* ctx, const Scalar& in, Scalar* out) {
-    checked_cast<BooleanScalar*>(out)->value = !in.is_valid;
+    auto options = OptionsWrapper<NanNullOptions>::Get(ctx);
+    bool* out_value = &checked_cast<BooleanScalar*>(out)->value;
+    if (in.is_valid) {
+      switch (in.type->id()) {
+        case Type::FLOAT:
+          *out_value = options.nan_is_null &&
+                       std::isnan(internal::UnboxScalar<FloatType>::Unbox(in));
+          break;
+        case Type::DOUBLE:
+          *out_value = options.nan_is_null &&
+                       std::isnan(internal::UnboxScalar<DoubleType>::Unbox(in));
+          break;
+        default:
+          *out_value = false;
+          break;
```

Comment:
Remove the `break` statement for the `default` case.

Reply:
Removed on 9a69e3e